### DISCUSSION TITLE
can ble.sh highlight opposing brace/paren pairs on hover like vim (with some setting I forget right now) does?

### DISCUSSION THREAD

**Discussion by bkerin** [akinomyoga/ble.sh#308] [General]
I see it supports % and recall reading that it parses bash syntax so seems possible, sorry if I missed it in the docs

**Comment by akinomyoga** [comment]
This is also in my todo list for a long time, but what is blocking it is that it requires the full parsing of the compound commands.

The current parser of ble.sh only analyzes Bash syntaxes that are sufficient to support the syntax highlighting and the syntax-aware completions, meaning that the nesting of `{` ... `}` or other compounds (such as `if` ... `fi`, `do` ... `done`, `case`...`esac`) are not tracked. The design choice is intended for efficiency. Bash is actually too slow to perform the full syntax analysis for every keystroke, so ble.sh implements the incremental parsing and tries to only update a part of the abstract syntax tree (AST) corresponding to the edited part and affected part around them. If the previous edit does not change the syntax structure, the updated subtree becomes minimal. However, if the previous edit caused structural changes, the affected range becomes large and we need to re-parse a larger subtree. Here, I wanted to make the AST as flat as possible because the full tracking of nested `{` ... `}`, etc. would make the structural change happen more frequently. This also applies to the counting of `(`...`)` in the arithmetic context.

- Now I think the recent versions of Bash is sufficiently fast, and the recent computers are also fast. So maybe we can perform the full-parsing of the nested compounds.
- Or we may keep the current syntax parsing, but backtrack and scan the command names for `{`...`}` to look up the corresponding brace afterward.
- Another way might be to highlight `{` and `}` regardless of the syntax structure, which seems to be the case in Vim. For example, consider `function a { echo }; }`. The first `}` is just an argument of `echo` and syntactically not the counterpart of the beginning `{`. But I wouldn't like to confuse the users by providing the wrong information, which is worse than providing nothing.

On the other hand, I think providing the paired highlighting for `${` ... `}`, `$(`...`)`, `$((`...`))`, etc. are relatively easier because they are parsed for the proper highlighting.

**Reply by akinomyoga** [reply]
I forgot to mention, but `%` is purely implemented in a simpler model of paired parens/braces/brackets without caring about the Bash syntax. But I would like to support the highlighting of matched paren based on the Bash syntax.

Also, for reference, Readline has a related feature with the setting `set blink-matching-paren on`, which `ble.sh` doesn't support now.

**Comment by bkerin** [comment]
Well first I agree if it risks slowing things down noticeably it's probably best to just avoid.  Latency effectively increases cognitive load and it's exactly when considering a complex/compound command that this is likely to be an issue.

But I also think ${} $() $(()) actually covers a good amount of the territory where this is most beneficial, because ble.sh already provides good support for multiline commands so hard commands can just be entered with good layout for function {} (subshell) etc to begin with.  ${} $() etc. are for interpolation so multiline layout is undesirable.  [] [[]] I guess aren't so easy (though it's tempting to think of using some heuristic for [[]] as it seems like a functionality superset of [] and unusual elsewhere).

**Reply by akinomyoga** [reply]
Thank you. Anyway, I'll reconsider implementing it in some way when I have time.
